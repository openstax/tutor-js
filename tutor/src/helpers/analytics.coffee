_ = require 'underscore'

Routes = require '../routes'

{CurrentUserStore} = require '../flux/current-user'

DestinationHelper  = require '../helpers/routes-and-destinations'

# generate custom event data for routes
Events =
  viewTaskStep: ({courseId}) ->
    # track all work done on a course
    Analytics.sendEvent 'Course', 'Work', label: courseId

  viewStudentDashboard: ({courseId}) ->
    # compare activity between courses
    Analytics.sendEvent 'Student', 'Dashboard', label: courseId


# Translators convert a url like '/foo/bar/123/baz/1' into a simplified one like just '/foo/bar'
Translators =

  dashboard:            ({courseId}) -> "/student/choose-course/#{courseId}"
  viewPractice:         ({courseId}) -> "/student/practice/#{courseId}"
  viewPerformanceForecast:  ({courseId}) -> "/student/performance-forecast/#{courseId}"
  viewStudentDashboard: ({courseId}) -> "/student/dashboard/#{courseId}"

  calendarByDate:        ({courseId}) -> "/teacher/calendar/#{courseId}"
  viewTeacherPerformanceForecast:      ({courseId}) -> "/teacher/performance-forecast/#{courseId}"
  viewScores:            ({courseId}) -> "/teacher/student-scores/#{courseId}"
  courseSettings:        ({courseId}) -> "/teacher/roster/#{courseId}"
  editReading:           ({courseId}) -> "/teacher/assignment/edit/reading/#{courseId}"
  editHomework:          ({courseId}) -> "/teacher/assignment/edit/homework/#{courseId}"
  editExternal:          ({courseId}) -> "/teacher/assignment/edit/external/#{courseId}"
  createReading:         ({courseId}) -> "/teacher/assignment/create/reading/#{courseId}"
  createHomework:        ({courseId}) -> "/teacher/assignment/create/homework/#{courseId}"
  createExternal:        ({courseId}) -> "/teacher/assignment/create/external/#{courseId}"
  calendarViewPlanStats: ({courseId}) -> "/teacher/metrics/quick/#{courseId}"
  reviewTask:            ({courseId}) -> "/teacher/metrics/review/#{courseId}"

  viewReferenceBook:        ({courseId}) -> "/reference-view/#{courseId}"
  viewReferenceBookSection: ({courseId}) -> "/reference-view/#{courseId}"
  viewReferenceBookPage:    ({courseId}) -> "/reference-view/#{courseId}"

  # Task steps are viewed by both teacher and student with no difference in params
  viewTaskStep:         ({courseId}) ->
    role = CurrentUserStore.getCourseRole(courseId, true)
    "/#{role}/task-step/#{courseId}"


GA = undefined

Analytics =

  setTracker: (tracker) -> GA = tracker

  sendPageView: (url) ->
    GA?('send', 'pageview', url)

  onNavigation: (path) ->
    return unless GA

    route = Routes.pathToEntry(path)

    return @sendPageView("/not-found/#{change.path}") unless route

    path = Translators[route.entry.name]?( route.match.params ) or route.match.pathname

    # if we're also going to send custom events then we set the page
    if Events[route.entry.name]
      GA('set', 'page', path)
      Events[route.entry.name]( route.match.params )
      @sendPageView() # url's not needed since it was set before events
    else
      @sendPageView(route.match.pathname)

  sendEvent: (category, action, attrs) ->
    return unless GA
    GA('send', 'event', category, action, attrs.label, attrs.value)

module.exports = Analytics
