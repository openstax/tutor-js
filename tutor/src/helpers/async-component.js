import React from 'react';
import AsyncErrorBoundary from '../components/error-monitoring/async-load-error';
import LoadingAnimation from '../../../shared/src/components/loading-animation'

export function asyncComponent(loader, name = '') {

    const Component = React.lazy(loader);
    const loading = <LoadingAnimation message={`Loading ${name}â€¦`} />;

    const Loader = (props) => (
        <AsyncErrorBoundary>
            <React.Suspense fallback={loading}>
                <Component {...props} />
            </React.Suspense>
        </AsyncErrorBoundary>
    );

    return (props) => <Loader {...props} />;
}


export function loadAsync(resolve, name) {
    // return a function so the router will only evaluate it when it's needed
    // in the future we may insert role dependant logic here
    return () => asyncComponent(resolve, name);
}
