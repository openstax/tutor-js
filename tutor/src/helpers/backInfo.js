import { navHistory } from '../models'


const backInfo = () => {
    // Gets route to last path that was not the same as the current one
    // See TransitionStore for more detail.
    const historyInfo = navHistory.latest

    const text = historyInfo != null && !!historyInfo.name ? `Back to ${historyInfo.name}` : 'My Courses';
    const to =  historyInfo != null && !!historyInfo.path ? historyInfo.path : 'myCourses';

    return {
        text,
        to,
    };
};


export default backInfo;
