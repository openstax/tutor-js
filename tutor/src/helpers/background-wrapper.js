import { styled } from 'vendor';
import { navbars } from '../theme.js';

const BackgroundWrapper = styled.div`
  background: #fff;
  min-height: calc(100vh - ${navbars.top.height} - ${navbars.bottom.height});
  position: relative;
  overflow: hidden;
  padding: 0 2.4rem;
`;

const ContentWrapper = styled.div`
  max-width: 1200px;
  margin: 0 auto;
`;

export { BackgroundWrapper, ContentWrapper };
