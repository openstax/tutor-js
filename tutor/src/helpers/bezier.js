function EASE_IN_OUT(t) {
    if (t < .5) { return 4 * t * t * t; } else { return ((t - 1) * ((2 * t) - 2) * ((2 * t) - 2)) + 1; }
}

function CALC(start, end, elapsed, duration) {
    if (elapsed > duration) { return end; }
    return start + ((end - start) * EASE_IN_OUT(elapsed / duration));
}

export default function bezier({
    range, duration, onStep, onComplete,
}) {
    const startTime = Date.now();
    const requestAnimationFrame = window.requestAnimationFrame || window.setTimeout;

    const step = () => {
        const elapsed = Date.now() - startTime;
        onStep(CALC(range[0], range[1], elapsed, duration));
        if (elapsed < duration) {
            requestAnimationFrame(step);
        } else if (onComplete) {
            onComplete();
        }
    };

    step();
}
