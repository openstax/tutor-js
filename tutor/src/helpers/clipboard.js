import _ from 'underscore';

const SUPPORTED = typeof document.queryCommandSupported === 'function' ? document.queryCommandSupported('copy') : undefined;

export default {
    isSupported: _.constant(!!SUPPORTED),

    copy() {
        if (!SUPPORTED) { return; }

        try {
            document.execCommand('copy');
        } catch (e) {
            // eslint-disable-next-line no-console
            console.warn('clipboard copy failed', e);
        }
    },
};
