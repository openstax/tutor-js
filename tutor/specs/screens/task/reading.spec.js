import UX from '../../../src/screens/task/ux';
import Reading from '../../../src/screens/task/reading';
import { TestRouter, Factory, FakeWindow, ld, TimeMock, C } from '../../helpers';

describe('Reading Tasks Screen', () => {
  let props, history;
  TimeMock.setTo('2017-10-14T12:00:00.000Z');

  beforeEach(() => {
    const task = Factory.studentTask({ type: 'reading' });
    history = new TestRouter({
      push: (url) => {
        props.ux.goToStep(Number(ld.last(url.split('/'))) - 1, false);
      },
    }).history;
    props = {
      windowImpl: new FakeWindow(),
      ux: new UX({ task, history, course: Factory.course() }),
    };
    history = new TestRouter({
      push: (url) => {
        props.ux.goToStep(ld.last(url.split('/')), false);
      },
    }).history;
  });

  it('matches snapshot', () => {
    expect(<C><Reading {...props} /></C>).toMatchSnapshot();
  });

  it('render as loading', () => {
    props.ux.goToStep(3);
    props.ux.currentStep.api.reset();
    expect(props.ux.currentStep.needsFetched).toBeTruthy();
    const r = mount(<C><Reading {...props} /></C>);
    expect(r).toHaveRendered('ContentLoader');
    r.unmount();
  });

  it('renders content', () => {
    const r = mount(<C><Reading {...props} /></C>);

    r.unmount();
  });

  it('renders value props', () => {
    props.ux.task.steps[1].formats = ['free-response', 'multiple-choice'];
    props.ux._stepIndex = props.ux.steps.findIndex(s=>s.type === 'two-step-intro');
    const r = mount(<C><Reading {...props} /></C>);
    expect(props.ux.currentStep.type).toEqual('two-step-intro');
    expect(r).toHaveRendered('TwoStepValueProp');
    r.unmount();
  });

  it('pages through steps', () => {
    const pr = mount(<C><Reading {...props} /></C>);
    expect(pr).toHaveRendered('a.paging-control.prev[disabled=true]');
    expect(pr).toHaveRendered('a.paging-control.next[disabled=false]');
    pr.find('a.paging-control.next').simulate('click');
    pr.find('a.paging-control.next').simulate('click');
    pr.unmount();
  });

});
