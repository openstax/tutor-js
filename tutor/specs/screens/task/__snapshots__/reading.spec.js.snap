// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reading Tasks Screen matches snapshot 1`] = `
<C>
  <ReadingTask
    ux={
      TaskUX {
        "_stepIndex": 0,
        "_task": StudentTask {
          "complete": undefined,
          "description": undefined,
          "due_at": 2017-10-17T12:00:00.000Z,
          "feedback_at": undefined,
          "id": 1,
          "isLoading": false,
          "is_deleted": undefined,
          "spy": Object {},
          "steps": Array [
            StudentTaskStep {
              "$nonSerializable": Object {
                "task": true,
              },
              "answer_id": undefined,
              "content": StudentTaskInteractiveStep {
                "chapter_section": Array [
                  1,
                  2,
                ],
                "has_learning_objectives": false,
                "html": "<span class=\\"os-text\\">Physics: An Introduction</span> </h2> <div class=\\"os-figure\\"> <figure id=\\"import-auto-id1580373\\"><span data-alt=\\"Two Canada geese flying close to each other in the sky.\\" data-type=\\"media\\" id=\\"import-auto-id1936836\\"><img alt=\\"Two Canada geese flying close to each other in the sky.\\" <div class=\\"os-caption-container\\"> <span class=\\"os-title-label\\">Figure </span><span class=\\"os-number\\">1.15</span><span class=\\"os-divider\\"> </span><span class=\\"os-caption\\"> <a class=\\"os-interactive-link\\" href=\\"https://www.openstaxcollege.org/l/02equation_grapher\\"> Click here for the simulation. </a> </span> </div> </div> </div></div> </div>",
                "id": 1,
                "related_content": Array [
                  Object {
                    "book_location": Array [
                      1,
                      2,
                    ],
                    "title": "Simple Machines",
                  },
                ],
                "title": "Electric Hazards and the Human Body",
                "type": "interactive",
              },
              "correct_answer_id": undefined,
              "external_url": null,
              "feedback_html": undefined,
              "formats": Array [],
              "free_response": undefined,
              "group": undefined,
              "id": 1,
              "isFetched": true,
              "is_completed": false,
              "labels": Array [],
              "multiPartGroup": undefined,
              "preview": "Simple Machines",
              "response_validation": Object {},
              "spy": Object {},
              "task": [Circular],
              "task_association_name": "steps",
              "type": "interactive",
              "uid": null,
            },
            StudentTaskStep {
              "$nonSerializable": Object {
                "task": true,
              },
              "answer_id": undefined,
              "content": StudentTaskExerciseStep {
                "_page": undefined,
                "average_step_number": undefined,
                "book": undefined,
                "content": Exercise {
                  "$nonSerializable": Object {
                    "wrapper": true,
                  },
                  "attachments": Array [],
                  "authors": Array [],
                  "copyright_holders": Array [],
                  "id": undefined,
                  "is_vocab": undefined,
                  "nickname": undefined,
                  "number": 11945,
                  "published_at": undefined,
                  "questions": Array [
                    ExerciseQuestion {
                      "$nonSerializable": Object {
                        "exercise": true,
                      },
                      "answers": Array [
                        ExerciseAnswer {
                          "$nonSerializable": Object {
                            "question": true,
                          },
                          "content_html": "geriatrics",
                          "correctness": undefined,
                          "feedback_html": undefined,
                          "id": "259301",
                          "question": [Circular],
                          "question_association_name": "answers",
                        },
                        ExerciseAnswer {
                          "$nonSerializable": Object {
                            "question": true,
                          },
                          "content_html": "gerontology",
                          "correctness": undefined,
                          "feedback_html": undefined,
                          "id": "259302",
                          "question": [Circular],
                          "question_association_name": "answers",
                        },
                        ExerciseAnswer {
                          "$nonSerializable": Object {
                            "question": true,
                          },
                          "content_html": "secondary aging",
                          "correctness": undefined,
                          "feedback_html": undefined,
                          "id": "259303",
                          "question": [Circular],
                          "question_association_name": "answers",
                        },
                        ExerciseAnswer {
                          "$nonSerializable": Object {
                            "question": true,
                          },
                          "content_html": "primary aging",
                          "correctness": undefined,
                          "feedback_html": undefined,
                          "id": "259304",
                          "question": [Circular],
                          "question_association_name": "answers",
                        },
                      ],
                      "collaborator_solutions": Array [],
                      "exercise": [Circular],
                      "exercise_association_name": "questions",
                      "formats": Array [
                        ExerciseFormat {
                          "$nonSerializable": Object {
                            "question": true,
                          },
                          "_format": "free-response",
                          "question": [Circular],
                          "question_association_name": "formats",
                        },
                        ExerciseFormat {
                          "$nonSerializable": Object {
                            "question": true,
                          },
                          "_format": "multiple-choice",
                          "question": [Circular],
                          "question_association_name": "formats",
                        },
                      ],
                      "hints": Array [],
                      "id": "62902",
                      "is_answer_order_important": false,
                      "stem_html": "In the early 1900s, a New York physician named Dr. Ignatz Nascher developed a medical specialty that focused on the elderly. This field is known broadly as ___.",
                      "stimulus_html": "",
                    },
                  ],
                  "stimulus_html": undefined,
                  "tags": Array [
                    ExerciseTag {
                      "raw": "time:short",
                    },
                    ExerciseTag {
                      "raw": "dok:2",
                    },
                    ExerciseTag {
                      "raw": "blooms:2",
                    },
                    ExerciseTag {
                      "raw": "type:practice",
                    },
                    ExerciseTag {
                      "raw": "requires-context:true",
                    },
                    ExerciseTag {
                      "raw": "book:stax-soc",
                    },
                    ExerciseTag {
                      "raw": "lo:stax-soc:13-2-2",
                    },
                    ExerciseTag {
                      "raw": "context-cnxmod:d9df5e48-4b72-482e-b616-886926188054",
                    },
                    ExerciseTag {
                      "raw": "context-cnxfeature:fs-id2332825",
                    },
                  ],
                  "uid": "11945@4",
                  "uuid": "7ddae47d-c4ec-437e-80bd-3f54d8a639a7",
                  "version": 4,
                  "versions": Array [],
                  "wrapper": [Circular],
                  "wrapper_association_name": "content",
                },
                "context": "<div data-type=\\"note\\" data-has-label=\\"true\\" id=\\"fs-id2332825\\" class=\\"sociological-research\\" a-type=\\"term\\" id=\\"import-auto-id1169033058271\\">geriatrics</span>, a medical specialty that focuses on the elderly. He created the word by combining were before Nascherâ€™s ideas gained acceptance?</p> </div></div> </div>",
                "course": undefined,
                "ecosystem_id": undefined,
                "has_interactive": false,
                "has_video": false,
                "id": 1,
                "isSelected": false,
                "is_excluded": false,
                "page_uuid": false,
                "pool_types": Array [],
                "preview": undefined,
                "question_stats": Array [],
                "related_content": Array [
                  RelatedContent {
                    "chapter_section": ChapterSection {
                      "chapter": 1,
                      "section": 2,
                    },
                    "page_id": undefined,
                    "title": "Simple Machines",
                    "uuid": undefined,
                  },
                ],
                "tags": Array [],
                "url": "",
              },
              "correct_answer_id": undefined,
              "external_url": null,
              "feedback_html": undefined,
              "formats": Array [
                "free-response",
              ],
              "free_response": undefined,
              "group": undefined,
              "id": 1,
              "isFetched": true,
              "is_completed": false,
              "labels": Array [],
              "multiPartGroup": undefined,
              "preview": "interactive: Prokaryotic Transcription",
              "response_validation": Object {},
              "spy": Object {},
              "task": [Circular],
              "task_association_name": "steps",
              "type": "exercise",
              "uid": "2@1",
            },
            StudentTaskStep {
              "$nonSerializable": Object {
                "task": true,
              },
              "answer_id": undefined,
              "content": StudentTaskInteractiveStep {
                "chapter_section": Array [
                  1,
                  2,
                ],
                "has_learning_objectives": false,
                "html": "<span class=\\"os-text\\">Physics: An Introduction</span> </h2> <div class=\\"os-figure\\"> <figure id=\\"import-auto-id1580373\\"><span data-alt=\\"Two Canada geese flying close to each other in the sky.\\" data-type=\\"media\\" id=\\"import-auto-id1936836\\"><img alt=\\"Two Canada geese flying close to each other in the sky.\\" <div class=\\"os-caption-container\\"> <span class=\\"os-title-label\\">Figure </span><span class=\\"os-number\\">1.15</span><span class=\\"os-divider\\"> </span><span class=\\"os-caption\\"> <a class=\\"os-interactive-link\\" href=\\"https://www.openstaxcollege.org/l/02equation_grapher\\"> Click here for the simulation. </a> </span> </div> </div> </div></div> </div>",
                "id": 2,
                "related_content": Array [
                  Object {
                    "book_location": Array [
                      1,
                      2,
                    ],
                    "title": "Electric Hazards and the Human Body",
                  },
                ],
                "title": "Electric Hazards and the Human Body",
                "type": "interactive",
              },
              "correct_answer_id": undefined,
              "external_url": null,
              "feedback_html": undefined,
              "formats": Array [],
              "free_response": undefined,
              "group": undefined,
              "id": 2,
              "isFetched": true,
              "is_completed": false,
              "labels": Array [],
              "multiPartGroup": undefined,
              "preview": "Food Irradiation",
              "response_validation": Object {},
              "spy": Object {},
              "task": [Circular],
              "task_association_name": "steps",
              "type": "interactive",
              "uid": null,
            },
            StudentTaskStep {
              "$nonSerializable": Object {
                "task": true,
              },
              "answer_id": undefined,
              "content": StudentTaskInteractiveStep {
                "chapter_section": Array [
                  1,
                  2,
                ],
                "has_learning_objectives": false,
                "html": "<span class=\\"os-text\\">Physics: An Introduction</span> </h2> <div class=\\"os-figure\\"> <figure id=\\"import-auto-id1580373\\"><span data-alt=\\"Two Canada geese flying close to each other in the sky.\\" data-type=\\"media\\" id=\\"import-auto-id1936836\\"><img alt=\\"Two Canada geese flying close to each other in the sky.\\" <div class=\\"os-caption-container\\"> <span class=\\"os-title-label\\">Figure </span><span class=\\"os-number\\">1.15</span><span class=\\"os-divider\\"> </span><span class=\\"os-caption\\"> <a class=\\"os-interactive-link\\" href=\\"https://www.openstaxcollege.org/l/02equation_grapher\\"> Click here for the simulation. </a> </span> </div> </div> </div></div> </div>",
                "id": 3,
                "related_content": Array [
                  Object {
                    "book_location": Array [
                      1,
                      2,
                    ],
                    "title": "Satellites and Keplerâ€™s Laws: An Argument for Simplicity",
                  },
                ],
                "title": "Electric Hazards and the Human Body",
                "type": "interactive",
              },
              "correct_answer_id": undefined,
              "external_url": null,
              "feedback_html": undefined,
              "formats": Array [],
              "free_response": undefined,
              "group": undefined,
              "id": 3,
              "isFetched": true,
              "is_completed": false,
              "labels": Array [],
              "multiPartGroup": undefined,
              "preview": "Angiosperms",
              "response_validation": Object {},
              "spy": Object {},
              "task": [Circular],
              "task_association_name": "steps",
              "type": "interactive",
              "uid": null,
            },
            StudentTaskStep {
              "$nonSerializable": Object {
                "task": true,
              },
              "answer_id": undefined,
              "content": StudentTaskExerciseStep {
                "_page": undefined,
                "average_step_number": undefined,
                "book": undefined,
                "content": Exercise {
                  "$nonSerializable": Object {
                    "wrapper": true,
                  },
                  "attachments": Array [],
                  "authors": Array [],
                  "copyright_holders": Array [],
                  "id": undefined,
                  "is_vocab": undefined,
                  "nickname": undefined,
                  "number": 11945,
                  "published_at": undefined,
                  "questions": Array [
                    ExerciseQuestion {
                      "$nonSerializable": Object {
                        "exercise": true,
                      },
                      "answers": Array [
                        ExerciseAnswer {
                          "$nonSerializable": Object {
                            "question": true,
                          },
                          "content_html": "geriatrics",
                          "correctness": undefined,
                          "feedback_html": undefined,
                          "id": "259301",
                          "question": [Circular],
                          "question_association_name": "answers",
                        },
                        ExerciseAnswer {
                          "$nonSerializable": Object {
                            "question": true,
                          },
                          "content_html": "gerontology",
                          "correctness": undefined,
                          "feedback_html": undefined,
                          "id": "259302",
                          "question": [Circular],
                          "question_association_name": "answers",
                        },
                        ExerciseAnswer {
                          "$nonSerializable": Object {
                            "question": true,
                          },
                          "content_html": "secondary aging",
                          "correctness": undefined,
                          "feedback_html": undefined,
                          "id": "259303",
                          "question": [Circular],
                          "question_association_name": "answers",
                        },
                        ExerciseAnswer {
                          "$nonSerializable": Object {
                            "question": true,
                          },
                          "content_html": "primary aging",
                          "correctness": undefined,
                          "feedback_html": undefined,
                          "id": "259304",
                          "question": [Circular],
                          "question_association_name": "answers",
                        },
                      ],
                      "collaborator_solutions": Array [],
                      "exercise": [Circular],
                      "exercise_association_name": "questions",
                      "formats": Array [
                        ExerciseFormat {
                          "$nonSerializable": Object {
                            "question": true,
                          },
                          "_format": "free-response",
                          "question": [Circular],
                          "question_association_name": "formats",
                        },
                        ExerciseFormat {
                          "$nonSerializable": Object {
                            "question": true,
                          },
                          "_format": "multiple-choice",
                          "question": [Circular],
                          "question_association_name": "formats",
                        },
                      ],
                      "hints": Array [],
                      "id": "62902",
                      "is_answer_order_important": false,
                      "stem_html": "In the early 1900s, a New York physician named Dr. Ignatz Nascher developed a medical specialty that focused on the elderly. This field is known broadly as ___.",
                      "stimulus_html": "",
                    },
                  ],
                  "stimulus_html": undefined,
                  "tags": Array [
                    ExerciseTag {
                      "raw": "time:short",
                    },
                    ExerciseTag {
                      "raw": "dok:2",
                    },
                    ExerciseTag {
                      "raw": "blooms:2",
                    },
                    ExerciseTag {
                      "raw": "type:practice",
                    },
                    ExerciseTag {
                      "raw": "requires-context:true",
                    },
                    ExerciseTag {
                      "raw": "book:stax-soc",
                    },
                    ExerciseTag {
                      "raw": "lo:stax-soc:13-2-2",
                    },
                    ExerciseTag {
                      "raw": "context-cnxmod:d9df5e48-4b72-482e-b616-886926188054",
                    },
                    ExerciseTag {
                      "raw": "context-cnxfeature:fs-id2332825",
                    },
                  ],
                  "uid": "11945@4",
                  "uuid": "7ddae47d-c4ec-437e-80bd-3f54d8a639a7",
                  "version": 4,
                  "versions": Array [],
                  "wrapper": [Circular],
                  "wrapper_association_name": "content",
                },
                "context": "<div data-type=\\"note\\" data-has-label=\\"true\\" id=\\"fs-id2332825\\" class=\\"sociological-research\\" a-type=\\"term\\" id=\\"import-auto-id1169033058271\\">geriatrics</span>, a medical specialty that focuses on the elderly. He created the word by combining were before Nascherâ€™s ideas gained acceptance?</p> </div></div> </div>",
                "course": undefined,
                "ecosystem_id": undefined,
                "has_interactive": false,
                "has_video": false,
                "id": 2,
                "isSelected": false,
                "is_excluded": false,
                "page_uuid": false,
                "pool_types": Array [],
                "preview": undefined,
                "question_stats": Array [],
                "related_content": Array [
                  RelatedContent {
                    "chapter_section": ChapterSection {
                      "chapter": 1,
                      "section": 2,
                    },
                    "page_id": undefined,
                    "title": "The Onset of Turbulence",
                    "uuid": undefined,
                  },
                ],
                "tags": Array [],
                "url": "",
              },
              "correct_answer_id": undefined,
              "external_url": null,
              "feedback_html": undefined,
              "formats": Array [
                "multiple-choice",
              ],
              "free_response": undefined,
              "group": undefined,
              "id": 2,
              "isFetched": true,
              "is_completed": false,
              "labels": Array [],
              "multiPartGroup": undefined,
              "preview": "one-to-one: RL Circuits",
              "response_validation": Object {},
              "spy": Object {},
              "task": [Circular],
              "task_association_name": "steps",
              "type": "exercise",
              "uid": "5@1",
            },
            StudentTaskStep {
              "$nonSerializable": Object {
                "task": true,
              },
              "answer_id": undefined,
              "content": StudentTaskReadingStep {
                "chapter_section": Array [
                  1,
                  2,
                ],
                "has_learning_objectives": true,
                "html": "<span class=\\"os-text\\">Physics: An Introduction</span> </h2> <div class=\\"os-figure\\"> <figure id=\\"import-auto-id1580373\\"><span data-alt=\\"Two Canada geese flying close to each other in the sky.\\" data-type=\\"media\\" id=\\"import-auto-id1936836\\"><img alt=\\"Two Canada geese flying close to each other in the sky.\\" data-media-type=\\"image/wmf\\" id=\\"90580\\" src=\\"https://cnx.org/resources/18f7913be5bf22af357a34143dd8638511b9ba9a\\" width=\\"325\\"></span> </figure><div class=\\"os-caption-container\\"> <span class=\\"os-title-label\\">Figure </span><span class=\\"os-number\\">1.2</span><span class=\\"os-divider\\"s as food calories, batteries, heat, light, and watch springs. Understanding this law makes it easier to learn about the various forms energy takes and how they relate to one another. Apparently unrelated topics are connected through broadly applicable physical laws, permitting an understanding beyond just the memorization of lists of facts.</p> <p id=\\"import-auto-id2555094\\">The unifying aspect of physical laws and the basic simplicity",
                "id": 1,
                "related_content": Array [
                  RelatedContent {
                    "chapter_section": ChapterSection {
                      "chapter": 1,
                      "section": 2,
                    },
                    "page_id": undefined,
                    "title": "Angiosperms",
                    "uuid": undefined,
                  },
                ],
                "title": "Simple Machines",
                "type": "reading",
              },
              "correct_answer_id": undefined,
              "external_url": null,
              "feedback_html": undefined,
              "formats": Array [],
              "free_response": undefined,
              "group": undefined,
              "id": 1,
              "isFetched": true,
              "is_completed": true,
              "labels": Array [],
              "multiPartGroup": undefined,
              "preview": "Environmental Limits to Population Growth",
              "response_validation": Object {},
              "spy": Object {},
              "task": [Circular],
              "task_association_name": "steps",
              "type": "reading",
              "uid": null,
            },
            StudentTaskStep {
              "$nonSerializable": Object {
                "task": true,
              },
              "answer_id": undefined,
              "content": StudentTaskInteractiveStep {
                "chapter_section": Array [
                  1,
                  2,
                ],
                "has_learning_objectives": true,
                "html": "<span class=\\"os-text\\">Physics: An Introduction</span> </h2> <div class=\\"os-figure\\"> <figure id=\\"import-auto-id1580373\\"><span data-alt=\\"Two Canada geese flying close to each other in the sky.\\" data-type=\\"media\\" id=\\"import-auto-id1936836\\"><img alt=\\"Two Canada geese flying close to each other in the sky.\\" <div class=\\"os-caption-container\\"> <span class=\\"os-title-label\\">Figure </span><span class=\\"os-number\\">1.15</span><span class=\\"os-divider\\"> </span><span class=\\"os-caption\\"> <a class=\\"os-interactive-link\\" href=\\"https://www.openstaxcollege.org/l/02equation_grapher\\"> Click here for the simulation. </a> </span> </div> </div> </div></div> </div>",
                "id": 4,
                "related_content": Array [
                  Object {
                    "book_location": Array [
                      1,
                      2,
                    ],
                    "title": "Simple Machines",
                  },
                ],
                "title": "Electric Hazards and the Human Body",
                "type": "interactive",
              },
              "correct_answer_id": undefined,
              "external_url": null,
              "feedback_html": undefined,
              "formats": Array [],
              "free_response": undefined,
              "group": undefined,
              "id": 4,
              "isFetched": true,
              "is_completed": false,
              "labels": Array [],
              "multiPartGroup": undefined,
              "preview": "Breathing",
              "response_validation": Object {},
              "spy": Object {},
              "task": [Circular],
              "task_association_name": "steps",
              "type": "interactive",
              "uid": null,
            },
          ],
          "students": Array [],
          "tasksMap": Object {},
          "title": "asymmetric",
          "type": "reading",
        },
        "course": Course {
          "appearance_code": "college_physics",
          "book_pdf_url": "https://archive.cnx.org/exports/185cbf87-c72e-48f5-b51e-f14f21b5eabd.pdf",
          "cloned_from_id": null,
          "current_role_id": 1,
          "default_due_time": "07:00",
          "default_open_time": "00:32",
          "does_cost": false,
          "ecosystem_book_uuid": "08df2bee-3db4-4243-bd76-ee032da173e8",
          "ecosystem_id": 1,
          "ends_at": "2017-11-14T13:00:00.000Z",
          "homework_progress_weight": 0.1,
          "homework_score_weight": 0.4,
          "id": 1,
          "is_access_switchable": true,
          "is_active": true,
          "is_college": true,
          "is_concept_coach": false,
          "is_lms_enabled": undefined,
          "is_lms_enabling_allowed": false,
          "is_preview": false,
          "just_created": false,
          "map": undefined,
          "name": "College Physics",
          "offering_id": 1,
          "periods": Array [
            CoursePeriod {
              "$nonSerializable": Object {
                "course": true,
              },
              "course": [Circular],
              "course_association_name": "periods",
              "default_due_time": "07:00",
              "default_open_time": "00:01",
              "enrollment_code": "66861",
              "enrollment_url": "http://localhost:3001/enroll/66861",
              "id": 1,
              "is_archived": false,
              "name": "1st",
              "num_enrolled_students": 46,
              "teacher_student_role_id": 1,
            },
            CoursePeriod {
              "$nonSerializable": Object {
                "course": true,
              },
              "course": [Circular],
              "course_association_name": "periods",
              "default_due_time": "07:00",
              "default_open_time": "00:01",
              "enrollment_code": "66861",
              "enrollment_url": "http://localhost:3001/enroll/66861",
              "id": 2,
              "is_archived": false,
              "name": "1st",
              "num_enrolled_students": 37,
              "teacher_student_role_id": 2,
            },
            CoursePeriod {
              "$nonSerializable": Object {
                "course": true,
              },
              "course": [Circular],
              "course_association_name": "periods",
              "default_due_time": "07:00",
              "default_open_time": "00:01",
              "enrollment_code": "66861",
              "enrollment_url": "http://localhost:3001/enroll/66861",
              "id": 3,
              "is_archived": false,
              "name": "1st",
              "num_enrolled_students": 31,
              "teacher_student_role_id": 3,
            },
          ],
          "reading_progress_weight": 0.2,
          "reading_score_weight": 0.2,
          "roles": Array [
            CourseRole {
              "$nonSerializable": Object {
                "course": true,
              },
              "course": [Circular],
              "course_association_name": "roles",
              "id": 1,
              "joined_at": 2017-09-21T12:00:00.000Z,
              "period_id": undefined,
              "research_identifier": undefined,
              "type": "student",
            },
          ],
          "salesforce_book_name": "College Physics",
          "starts_at": "2017-09-14T12:00:00.000Z",
          "students": Array [
            CourseStudent {
              "$nonSerializable": Object {
                "course": true,
              },
              "course": [Circular],
              "course_association_name": "students",
              "first_name": "Adelia",
              "id": 1,
              "is_active": true,
              "is_comped": false,
              "is_paid": false,
              "is_refund_allowed": undefined,
              "is_refund_pending": undefined,
              "last_name": "Gutmann",
              "name": undefined,
              "payment_due_at": 2017-10-17T12:00:00.000Z,
              "period_id": undefined,
              "prompt_student_to_pay": false,
              "role_id": 1,
              "roster": undefined,
              "student_identifier": "",
              "uuid": undefined,
            },
          ],
          "term": "spring",
          "time_zone": "Central Time (US & Canada)",
          "uuid": "ebf88799-1457-46ad-a417-7f3869c17dc9",
          "webview_url": "https://qa.cnx.org/contents/185cbf87-c72e-48f5-b51e-f14f21b5eabd",
          "year": 2017,
        },
        "history": Object {
          "createHref": [MockFunction],
          "listen": [MockFunction],
          "location": Object {
            "pathname": "/",
          },
          "push": [Function],
          "replace": [MockFunction],
        },
        "isLocked": false,
        "viewedInfoSteps": Array [],
        "window": [Window],
      }
    }
    windowImpl={
      FakeWindow {
        "addEventListener": [MockFunction],
        "clearInterval": [MockFunction],
        "document": Object {
          "body": <body />,
          "hidden": false,
        },
        "getSelection": [MockFunction],
        "history": Object {
          "pushState": [MockFunction],
        },
        "innerHeight": 1024,
        "innerWidth": 768,
        "localStorage": Object {
          "getItem": [MockFunction],
          "setItem": [MockFunction],
        },
        "location": Object {
          "href": "http://localhost:3001/dashboard",
          "pathname": "/",
          "reload": [MockFunction],
          "search": "",
        },
        "open": [MockFunction],
        "pageXOffset": 0,
        "pageYOffset": 0,
        "querySelector": [MockFunction],
        "removeEventListener": [MockFunction],
        "requestAnimationFrame": [MockFunction],
        "screen": Object {
          "height": 1024,
          "width": 768,
        },
        "scroll": [MockFunction],
        "setInterval": [MockFunction],
      }
    }
  />
</C>
`;
